sky/spot/controller.py

```diff
@@ -188,15 +191,15 @@ def run(self):
             # The exception will be caught when:
             # 1. The strategy_executor fails to launch/recover the cluster
             #    after the max number of retries when retry_until_up is not set.
-            # 2. None the failovers are caused by resource unavailability, i.e.
-            #    the job is configured incorrectly, e.g.,
-            #    InvalidClusterNameError, NotSupportedError,
+            # 2. None of the failovers are caused by resource unavailability;
+            #    i.e., they are caused by errors before actual provisioning,
+            #    e.g., InvalidClusterNameError, NotSupportedError,
             #    CloudUserIdentityError, etc.
             logger.error(f'{common_utils.class_fullname(e.__class__)}: '
                          f'{colorama.Fore.RED}{e}{colorama.Style.RESET_ALL}')
-            if any(
+            if (not e.failover_history or any(
```

Comment:
Document when `not e.failover_history` will happen and why we consider that case to be FAILED_NO_RESOURCE?

Reply:
I refactored this part of the code for better readability. PTAL. : )